tests/xbot/common/logic/LatchTest.java

```diff
@@ -0,0 +1,133 @@
+package xbot.common.logic;
+
+import org.junit.Test;
+
+import java.util.Observable;
+import java.util.Observer;
+
+import static junit.framework.TestCase.assertEquals;
+
+public class LatchTest {
+
+    @Test
+    public void testRisingEdge() {
+        Latch latch = new Latch(false, Latch.EdgeType.RisingEdge);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), null);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+    }
+
+    @Test
+    public void testFallingEdge() {
+        Latch latch = new Latch(true, Latch.EdgeType.FallingEdge);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+    }
+
+    @Test
+    public void testRiseFallRiseObservingRisingEdge() {
+        Latch latch = new Latch(false, Latch.EdgeType.RisingEdge);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), null);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getNumTimesUpdated(), 2);
+    }
+
+    @Test
+    public void testFallRiseFallObservingFallingEdge() {
+        Latch latch = new Latch(true, Latch.EdgeType.FallingEdge);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
```

Comment:
Can we assert on the number of updates at each step (all tests)? Part of the contract of the latch is that it only fires an update when the edge actually happens; I'm slightly worried that two complementary bugs could cancel each other out or a logic inversion could be introduced and the tests would still pass.

Reply:
Added